usr/src/cmd/zdb/zdb.c

```diff
@@ -2301,24 +2301,29 @@ dump_label(const char *dev)
 
 		(void) snprintf(path, sizeof (path), "%s%s", ZFS_RDISK_ROOTD,
 		    dev);
-		if ((s = strrchr(dev, 's')) == NULL || !isdigit(*(s + 1)))
+		if (((s = strrchr(dev, 's')) == NULL &&
+		    (s = strchr(dev, 'p')) == NULL) ||
+		    !isdigit(*(s + 1)))
 			(void) strlcat(path, "s0", sizeof (path));
 	}
 
-	if (stat64(path, &statbuf) != 0) {
-		(void) printf("failed to stat '%s': %s\n", path,
+	if ((fd = open64(path, O_RDONLY)) < 0) {
+		(void) fprintf(stderr, "cannot open '%s': %s\n", path,
 		    strerror(errno));
 		exit(1);
 	}
 
```

Comment:
It would be nice to have any proof reference in comment why fstat64() is required here and not stat64(). POSIX and man page don't describe difference in behaviour between those functions for char devices:
    st_size
                     For regular files, this is the address of the end of the
                     file. For block special or character special, this is not
                     defined. See also pipe(2).

And therefore comment about using fstat64() here is strongly recommended.